database.hh

```diff
@@ -845,7 +845,8 @@ public:
      */
     future<std::unordered_set<sstring>> get_sstables_by_partition_key(const sstring& key) const;
 
-    const sstables::sstable_set& get_sstable_set() const;
+    sstables::sstable_set& get_sstable_set() const;
+    lw_shared_ptr<sstables::sstable_set> get_sstable_set_copy() const;
```

Comment:
Considering it caused confusion, it probably does need a comment. 
This method is needed for cases where we want to read the state of the sstable_set but also do modifications that don't affect the sstable_set in the table (there are two such cases: a copy is used for expiration purposes in compaction, and in table::make_reader_excluding_sstables() we give a modified sstable_set to an sstable_reader). This can't be solved without copying, because in our approach all versions are linear.
Also, it doesn't create a copy of all versions, just of the last one(all sstables that are alive in it)

Reply:
We might use copy-on-write (create-version-on-write) for this. 

What's the difficulty with non-linear versions?

Reply:
The biggest issues with keeping non-linear versions in the versioned sstable_set are that it would be harder to decide if a version is between two other versions - the versions now form a tree, so they can't really be represented by integers, and it would also be harder to decide when all references to an sstable die - the sstable might have been erased multiple times on different branches. We would also need to allow modifying all versions that are alive - not only the last one like we currently do.

Reply:
Suppose an entry in the data structure contains a `vector<version_status>` where `version_status` is `struct { version* v, bool included }`.

To copy a set, create a new version object and mark it as a descendent of the version object you copied from.
To add or remove an sstable, append a version_status with included set to true/false.

To check if an sstable is in the set, walk the vector backwards until you find a version that is equal to your version, or is an ancestor.

To delete a version, iterate over all related entries (you'll have to keep pointers to those entries) and replace them with pointers to the deleted version's descendants.

But I feel like we're reinventing a wheel here. Check purely functional data structures, they have to accomplish the same thing since they cannot mutate the data structure.

Another alternative is to check if every place we use sstable_set is strictly necessary. Of course, converting to sstable_list is also O(n), so we should only use sstable_list if we're sure n is small. Maybe we can remove enough users so that linear versions become workable.

Reply:
I realize that writing a purely functional data structure would basically solve all our problems, but I was under an impression that we are trying to use a simpler solution here.

In your example, am I missing something or do we still have to walk over all ancestors of a version to check if the vector has a pointer to one of them (there are ways to check it in log(n) but that would need additional solutions)? And what is the complexity of deleting a version here, won't we potentially iterate over all sstables in the set? I had a solution that doesn't need to iterate over all sstables related to the deleted version, but only if there was only one following version.

Reply:
> I realize that writing a purely functional data structure would basically solve all our problems, but I was under an impression that we are trying to use a simpler solution here.

Simpler is better! But if we end up just moving the expensive operations from one place to another, ore eliminating just half the expensive operations, it doesn't help

If the expensive operations become very rare, it's okay.

> 
> In your example, am I missing something or do we still have to walk over all ancestors of a version to check if the vector has a pointer to one of them (there are ways to check it in log(n) but that would need additional solutions)? And what is the complexity of deleting a version here, won't we potentially iterate over all sstables in the set?

If a version changed only a few sstables (quite common), and if the version has a list of those sstables, then it becomes a cheap operation (but indeed the complexity keep increasing). We might say that a version has a link to the parent version, plus a delta compared to the previous version.

> I had a solution that doesn't need to iterate over all sstables related to the deleted version, but only if there was only one following version.

It may still be possible. If we understand the case in compaction well enough, maybe we can fit it into the solution. 

I don't want this to grow into a huge research problem, but I also don't want to solve just half of it.

It's worth looking at the functional data structure literature. Maybe there is something that is easy to adapt. I expect most rely on garbage collection, but perhaps it can be replaced by reference counts or just intrusive lists.

Reply:
It looks like the only cases where we need to use non-linear versions (table::make_reader_excluding_sstables() and regular_compaction::update_pending_ranges()) don't need to create any more versions, because after each change they need only the state with all changes.
So what we can do is keep local changes on the non-linear version, and merge its results with the snapshots when selecting, similarly to the local changes in the sstable_list. 
With this we could completely avoid copying the sstable_set, but we would need to additionally remember the sstables that were erased in the non-linear version and filter them out while selecting. On the other hand, we would still need to do something similar, even if we were using a purely functional data structure for this. 
Actually, if we were erasing a large number of sstables, maybe it would be better to just copy the set?

Reply:
If I understand correctly, you are suggesting to replace

     s1 = main_set
     s1.remove(some_sstables)

with 

    s1 = main_set   # create a snapshot, but do not mutate it
    s1_delta = some_sstables

and follow by looking up both s1 and s1_delta?

but could be not remember some_sstables in the data structure directly?


Typically, sstable_set evolution does not affect many sstables. With Leveled Compaction Strategy we typically change ~22 sstables (1 sstable promoted, 11 sstables intersecting it, 10-11 new sstables).

With Incremental Compaction Strategy it's just a few each time.

@nyh / @psarna  can you comment on the expected size of excluded_sstables in view_updating_consumer?

Reply:
If versions were a single linked list, we could create a new version based on a snapshot, without adding nonlinear version support:
```
s1 = sst_set();
//.. modifications to s1
s2 = sst_set(s1.snapshot());
// modification to s2 just add versions on top of s1's snapshot, but those are not visible to s1
```

Reply:
Considering all arguments, I think versions with pointers to one another are a good idea. Assuming we have very few concurrently alive versions at any time, we can even drop the "versioned_sstables" that were stored in a single container and later filtered, and merge results from containers from only the relevant versions instead. 

With linked snapshots, we can quite easily support non-linear versions, at least in a way that isn't very efficientwhen there are too many of them, but I think that's not a big problem (still not close to copying everything). Actually, if we have all those pointers, we could probably also remove the versioned_sstable_set altogether. 

I'll also ask ahead because I often get this comment in review, would it be fine to have lw_shared_ptrs to parent versions? We need some shared owning because we may have a version that doesn't belong to any versioned_sstable_set, and I don't think we want to do the mallocing ourselves.

Reply:
We don't actually have anything against using `lw_shared_ptr<>` when its justified. It's only when they can be seen everywhere in a patchset that it starts to raise eyebrows and people start to question whether they are really necessary.

Reply:
Ah, so sstable_set's copy constructor will create a live snapshot.

There's no problem with lw_shared_ptr (apart that it's better not to use them when a value is fine). When changing a widely used API like this, they should be hidden in the implementation, not exposed.